runtime/j9vm/javanextvmi.cpp

```diff
@@ -681,11 +685,159 @@ JVM_TakeVirtualThreadListToUnblock(JNIEnv *env, jclass ignored)
+			if (NULL != carrierVMThread->currentContinuation) {
+				targetThread = carrierVMThread;
+			}
+		}
+		j9object_t continuationObj = J9VMJAVALANGVIRTUALTHREAD_CONT(currentThread, threadObject);
+		ContinuationState continuationState = *VM_ContinuationHelpers::getContinuationStateAddress(currentThread, continuationObj);
+		if ((JVMTI_VTHREAD_STATE_NEW != vthreadState)
+		&& (JVMTI_VTHREAD_STATE_TERMINATED != vthreadState)
+		&& !VM_ContinuationHelpers::isFinished(continuationState)
+		) {
+			*isThreadAlive = JNI_TRUE;
+		}
+	} else {
+		targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, threadObject);
+		*isThreadAlive = J9VMJAVALANGTHREAD_STARTED(currentThread, threadObject) && (NULL != targetThread);
+	}
+
+	if (!isThreadAlive) {
+		f_monitorExit(vm->vmThreadListMutex);
+		if (isVirtualThread) {
+			vm->internalVMFunctions->releaseVThreadInspector(currentThread, thread);
+		}
+		return NULL;
+	}
+	if (NULL != targetThread) {
+		targetThread->inspectorCount += 1;
+	}
+
+	f_monitorExit(vm->vmThreadListMutex);
+	return targetThread;
+}
+
+static void
+releaseTargetVMThread(J9VMThread *currentThread, J9VMThread *targetThread, jthread thread, jboolean isVirtualThread)
+{
+	j9object_t threadObject = J9_JNI_UNWRAP_REFERENCE(thread);
+	if ((currentThread->threadObject != threadObject) && isVirtualThread) {
+		currentThread->javaVM->internalVMFunctions->releaseVThreadInspector(currentThread, thread);
+	}
+	if (currentThread != targetThread) {
+		J9JavaVM *vm = targetThread->javaVM;
+		/* Release the J9VMThread (allow it to die) now that we are no longer inspecting it. */
+		f_monitorEnter(vm->vmThreadListMutex);
+		if (0 == --(targetThread->inspectorCount)) {
+			f_monitorNotifyAll(vm->vmThreadListMutex);
+		}
+		f_monitorExit(vm->vmThreadListMutex);
+	}
+}
+
 JNIEXPORT jobject JNICALL
 JVM_CreateThreadSnapshot(JNIEnv *env, jobject thread)
 {
-	Assert_SC_true(!"JVM_CreateThreadSnapshot unimplemented");
-	return NULL;
+	J9VMThread *currentThread = (J9VMThread *)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions const * const vmfns = vm->internalVMFunctions;
+	j9object_t resultObject = NULL;
+	jobject result = NULL;
+	vmfns->internalEnterVMFromJNI(currentThread);
+
+	if (NULL == thread) {
+		vmfns->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGINTERNALERROR, NULL);
+	} else {
+		const char* className = "jdk/internal/vm/ThreadSnapshot";
+		UDATA classNameLen = strlen(className);
+		J9Class* clazz = NULL;
+		J9ClassLoader *classLoader = vm->systemClassLoader;
+		j9object_t threadObject = J9_JNI_UNWRAP_REFERENCE(thread);
+		jboolean isVirtual = IS_JAVA_LANG_VIRTUALTHREAD(currentThread, threadObject);
+		jboolean isAlive = JNI_FALSE;
+		J9VMThread* targetThread = getTargetVMThread(currentThread, thread, isVirtual, &isAlive);
+
+		if (!isAlive) {
+			goto done;
+		}
+
+		clazz = vmfns->internalFindClassUTF8(currentThread, (U_8*)className, classNameLen, classLoader, J9_FINDCLASS_FLAG_EXISTING_ONLY);
```

Comment:
since you added a vmconstantpool entry, you can just get the class via `J9VMJDKINTERNALVM_THREADSNAPSHOT_OR_NULL`